### DISCUSSION TITLE
exposing the the flux command for embedding?

### DISCUSSION THREAD

**Discussion by ghost** [fluxcd/flux2#3215] [Proposals]
Simple making the changes I made [in this fork](https://github.com/protosam/flux2) to make it possible to embed the flux command [like kubectl can be embedded](https://github.com/kubernetes/kubectl/blob/92206a7303970ac055539a8ba6f93775d5f7643f/pkg/cmd/cmd.go#L96).

This makes it possible for people to vendor complete cli toolkits.

**Comment by kingdonb** [comment]
This is something we think we would consider in the scope of changes that might be possible for GA

We want to be careful about creating new public API surfaces because right now we have the flexibility to change how the CLI works without breaking downstream consumers (the stability guarantees are not there for the CLI internals as I understand our current position, we're maintaining the capability to break things in there until we've declared general availability, then we will add some new stability guarantees...)

More information about GA here: https://fluxcd.io/roadmap/#the-road-to-flux-v2-ga

I think based on our discussion at Flux Bug Scrub about this today, this is something we will consider, but it is going to require some thought and can't be done right away.

If you can tell us what you need it for with a bit of specificity, we can perhaps point you to the Flux functionality in the relevant package so you can use it from the APIs which are already exposed and considered as stable? For example, all of the new OCI functionality is supposed to be exposed in packages already.

**Reply by ghost** [reply]
The idea is to just expose something that already exists, the `*cobra.Cli`. Has zero impact down stream.

**Reply by stefanprodan** [reply]
The CLI binary embeds files which are generated at release time, having the manifests in flux2 repo is not what we want. The CLI is not designed to be used as a Go package nor it should be imported as such. We plan to extract the commands underlying logic to dedicated Go packages under `fluxcd/pkg` that you could use in your own Go project.

**Reply by ghost** [reply]
Oh! You're right, I forgot about the generated manifests. The CI would need to change. Pre-generated manifests would need to live in branch.

**Reply by devantler** [reply]
@stefanprodan Happy to find this thread! I am rebuilding [a CLI tool called ksail in Go](https://github.com/devantler-tech/ksail-go) with the aim of providing a tool that can both provision clusters, and manage workloads by both embedding tools like kind, k3d, k9s, helm, kubectl, flux, argocd etc. I now reached the point where I want to embed flux and its subcommands, but I saw that the Cobra commands are internal. The major motivation for rebuilding it in Go is that many tools can be embedded via native Go dependencies, and thus remove the need for users to have all the different binaries preinstalled.

Looking at the `fluxcd/pkg` I feel limited. I really would like to implement equivalents to `flux create|delete|get`, but preferably by relying on upstream Go packages. Do you want to do more towards supporting this, or might you have a suggestion to help me?
